File: Xenon/Xenon/AudioMenuState.ts
import CGameState = require("GameState");
import CApplication = require("Application");
import CScene = require("Scene");
import CStarfield = require("Starfield");
import gsCControls = require("Controls");
import enums = require("Enums");
import Options = require("Options");
import gsCPoint = require("Point");
import gsCMenu = require("Menu");
import COptionsMenuState = require("OptionsMenuState");

class CAudioMenuState extends CGameState {

    m_optionState: COptionsMenuState;

    constructor(scene?: CScene, starfield?: CStarfield, font8x8?: HTMLImageElement, font16x16?: HTMLImageElement, app?: CApplication, ctx?: CanvasRenderingContext2D, menuState?, menu?, optionState?: COptionsMenuState) {
        super(font8x8, font16x16, app, ctx);

        this.m_starfield = starfield;
        this.m_mainMenuState = menuState;
        this.m_menu = menu;
        this.m_optionState = optionState;

        this.m_stateName = "AudioMenuState";
    }

    //-------------------------------------------------------------

    public instance(): CGameState {
        return this.m_app.instance = this;
    }

    //-------------------------------------------------------------

    public copyOptionsToMenu(): void {
        this.m_menu.setValue(enums.AudioMenuItem.OM_MUSIC, this.m_options.getOption(enums.OptionType.OPTION_MUSIC));
        this.m_menu.setValue(enums.AudioMenuItem.OM_SOUNDFX, this.m_options.getOption(enums.OptionType.OPTION_SOUNDFX));
    }

    //-------------------------------------------------------------

    public copyMenuToOptions(): void {
        this.m_options.setOption(enums.OptionType.OPTION_MUSIC, this.m_menu.getValue(enums.AudioMenuItem.OM_MUSIC));
        this.m_options.setOption(enums.OptionType.OPTION_SOUNDFX, this.m_menu.getValue(enums.AudioMenuItem.OM_SOUNDFX));
    }

    //-------------------------------------------------------------

    public create(): boolean {
        this.m_menu.clear();
        this.m_menu.addSlider("   Music", 10, 0, 10);
        this.m_menu.addSlider("Sound FX", 10, 0, 10);
        this.copyOptionsToMenu();
        //this.m_menu.addSeperator();
        this.m_menu.addSelection("Apply");
        this.m_menu.addSelection("Cancel");
        this.m_menu.setWrap(true);
        this.m_menu.setPosition(new gsCPoint(0, 150));
        this.m_menu.setSpacing(new gsCPoint(0, 30));
        this.m_menu.setCurrentItem(enums.AudioMenuItem.OM_CANCEL);
        this.m_menu.setFont(this.m_medium_font);

        return true;
    }

    //-------------------------------------------------------------

    public update(ctx: CanvasRenderingContext2D, controls: gsCControls): boolean {

        //	if (!CGameState::update())
        //		return false;

        if (this.m_options.getOption(enums.OptionType.OPTION_BACKDROP)) {
            ctx.drawImage(this.backgroundTexture, 0, 0);
        }

        this.m_starfield.Update(4);
        this.m_starfield.Draw(ctx);
        this.m_medium_font.setTextCursor(new gsCPoint(0, 50));
        this.m_medium_font.justifyString("Audio Options");
        this.m_menu.draw(ctx);

        var item: enums.AudioMenuItem = <enums.AudioMenuItem>this.m_menu.getCurrentItem();

        if (controls.returnPressed || controls.enterPressed || controls.lcontrolPressed) {
            controls.returnPressed = false;
            controls.enterPressed = false;
            controls.lcontrolPressed = false;

            if (item == enums.AudioMenuItem.OM_APPLY) {
                super.playSample(enums.GameSampleType.SAMPLE_MENU_SELECTION);
                this.copyMenuToOptions();
                if (this.m_options.areChanged()) {
                    //gsCFile::setDirectory(DIRECTORY_ROOT);
                    //Options.save(OPTIONS_FILENAME);
                    //updateVolume();
                    //if (Options.requireReload()) {
                    //CMessageBoxState::setup("Restart to apply options ?",
                    //							"Yes",0,
                    //							"No",COptionsMenuState::instance(),
                    //							true);
                    //return changeState(CMessageBoxState::instance());
                    return;
                }
                else {
                    return this.changeState(this.m_app.instance = this.m_optionState);
                }
                //}
                //else {
                //    return this.changeState(this.m_app.instance = this.m_optionState);
                //}
            }
            else if (item == enums.AudioMenuItem.OM_CANCEL) {
                super.playSample(enums.GameSampleType.SAMPLE_MENU_BACK);
                return this.changeState(this.m_app.instance = this.m_optionState);
            }
        }

        if (controls.up) {
            super.playSample(enums.GameSampleType.SAMPLE_MENU_SELECTION);
            this.m_menu.scroll(-1);
        }
        if (controls.down) {
            super.playSample(enums.GameSampleType.SAMPLE_MENU_SELECTION);
            this.m_menu.scroll(1);
        }
        if (controls.left) {
            if (this.m_menu.setValue(item, this.m_menu.getValue(item) - 1)) {
                super.playSample(enums.GameSampleType.SAMPLE_MENU_OPTION);
            }
        }
        if (controls.right) {
            if (this.m_menu.setValue(item, this.m_menu.getValue(item) + 1)) {
                super.playSample(enums.GameSampleType.SAMPLE_MENU_OPTION);
            }
        }
        return true;
    }

    //-------------------------------------------------------------
}

export = CAudioMenuState;

File: Xenon-Original_C++_Code/xenon/source/audiomenustate.cpp
//-------------------------------------------------------------
//
// Class:	CAudioMenuState
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CGameState
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

CAudioMenuState *CAudioMenuState::m_instance = 0;

//-------------------------------------------------------------

CAudioMenuState::CAudioMenuState()
{
}

//-------------------------------------------------------------

CAudioMenuState::~CAudioMenuState()
{
}

//-------------------------------------------------------------

CGameState *CAudioMenuState::instance()
{
	if (!m_instance)
		m_instance = new CAudioMenuState;

	return m_instance;
}

//-------------------------------------------------------------

void CAudioMenuState::copyOptionsToMenu()
{
	m_menu.setValue(OM_MUSIC		,Options.getOption(OPTION_MUSIC));
	m_menu.setValue(OM_SOUNDFX		,Options.getOption(OPTION_SOUNDFX));
}

//-------------------------------------------------------------

void CAudioMenuState::copyMenuToOptions()
{
	Options.setOption(OPTION_MUSIC		,m_menu.getValue(OM_MUSIC));
	Options.setOption(OPTION_SOUNDFX	,m_menu.getValue(OM_SOUNDFX));
}

//-------------------------------------------------------------

bool CAudioMenuState::create()
{
	m_menu.clear();

	m_menu.addSlider("   Music",10,0,10);
	m_menu.addSlider("Sound FX",10,0,10);

	copyOptionsToMenu();

	m_menu.addSeperator();
	m_menu.addSelection("Apply");
	m_menu.addSelection("Cancel");

	m_menu.setWrap(true);
	m_menu.setPosition(gsCPoint(0,150));
	m_menu.setSpacing(gsCPoint(0,30));
	m_menu.setCurrentItem(OM_CANCEL);
	m_menu.setFont(&m_medium_font);

	return true;
}

//-------------------------------------------------------------

bool CAudioMenuState::update()
{
	if (!CGameState::update())
		return false;

	if (Options.getOption(OPTION_BACKDROP))
		m_backdrop.draw(gsCPoint(0,0));
	else
		m_screen.clear(gsCColour(gsBLACK));

	m_starfield.move(4);
	m_starfield.draw();

	m_medium_font.setTextCursor(gsCPoint(0,50));
	m_medium_font.justifyString("Audio Options");

	m_menu.draw();
	
	m_screen.flip();

	AudioMenuItem item = (AudioMenuItem) m_menu.getCurrentItem();

	switch (getKey()) {
		case gsKEY_RETURN:
		case gsKEY_ENTER:
		case gsKEY_LCONTROL:
			if (item == OM_APPLY) {
				CGameState::playSample(SAMPLE_MENU_SELECTION);
				copyMenuToOptions();
				if (Options.areChanged()) {

					gsCFile::setDirectory(DIRECTORY_ROOT);

					Options.save(OPTIONS_FILENAME);

					updateVolume();

					if (Options.requireReload()) {
						CMessageBoxState::setup("Restart to apply options ?",
												"Yes",0,
												"No",COptionsMenuState::instance(),
												true);
						return changeState(CMessageBoxState::instance());
						}
					else
						return changeState(COptionsMenuState::instance());
					}
				else
					return changeState(COptionsMenuState::instance());
				}
			else if (item == OM_CANCEL) {
				CGameState::playSample(SAMPLE_MENU_BACK);
				return changeState(COptionsMenuState::instance());
				}
			break;
		case gsKEY_UP:
			CGameState::playSample(SAMPLE_MENU_SELECTION);
			m_menu.scroll(-1);
			break;
		case gsKEY_DOWN:
			CGameState::playSample(SAMPLE_MENU_SELECTION);
			m_menu.scroll(1);
			break;
		case gsKEY_LEFT:
			if (m_menu.setValue(item,m_menu.getValue(item) - 1))
				CGameState::playSample(SAMPLE_MENU_OPTION);
			break;
		case gsKEY_RIGHT:
			if (m_menu.setValue(item,m_menu.getValue(item) + 1))
				CGameState::playSample(SAMPLE_MENU_OPTION);
			break;
		}

	return true;
}

//-------------------------------------------------------------

